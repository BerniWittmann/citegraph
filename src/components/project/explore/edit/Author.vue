<template>
  <div class="project-explore-single-author-edit">
    <v-card class="pa-4">
      <v-card-title class="headline">
        <h2 class="display-5" v-once>{{ author.firstName }} {{ author.lastName }}</h2>
      </v-card-title>
      <v-container>
        <v-form ref="form" @submit.prevent="submit" class="mt-12" v-model="valid">
          <v-row justify="center">
            <v-col lg="6">
              <v-text-field
                :label="$t('project.explore.edit.author.first_name')"
                v-model="firstName"
                :rules="nameRules"
                required
                autofocus/>
            </v-col>
          </v-row>
          <v-row justify="center">
            <v-col lg="6">
              <v-text-field
                :label="$t('project.explore.edit.author.last_name')"
                v-model="lastName"
                :rules="nameRules"
                required
                autofocus/>
            </v-col>
          </v-row>
          <v-row class="my-10" justify="space-between">
            <v-col cols="auto">
              <v-btn rounded large color="warning" @click="back">
                <v-icon>mdi-menu-left</v-icon>
                {{ $t('project.explore.edit.back') }}
              </v-btn>
            </v-col>

            <v-col cols="auto">
              <v-btn
                rounded large
                :disabled="!valid"
                :loading="loading"
                color="success"
                type="submit"
              >{{ $t('project.explore.edit.submit') }}
              </v-btn>
            </v-col>
          </v-row>
        </v-form>
      </v-container>
    </v-card>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import Author from '@/models/paperEntities/author'

@Component
export default class AuthorEditPage extends Vue {
  valid: boolean = true
  loading: boolean = false
  firstName: string = ''
  lastName: string = ''

  nameRules: Array<Function> = [
    (v: string | undefined) => !!v || this.$root.$t('project.explore.edit.author.name_required'),
    (v: string | undefined) => (v && v.trim().length > 0) || this.$root.$t('project.explore.edit.author.name_not_empty')
  ]

  urlRules: Array<Function> = []

  get author (): Author {
    return this.$store.getters['paperEntities/activeEntity'] as Author
  }

  beforeMount (): void {
    this.firstName = this.author.firstName
    this.lastName = this.author.lastName
  }

  async submit (): Promise<void> {
    if (!this.valid) return
    this.loading = true
    await this.$store.dispatch('paperEntities/updateEntity', {
      params: {
        projectId: this.$route.params.projectId,
        entityType: this.$route.params.queryByType,
        id: this.$route.params.entityId
      },
      data: {
        firstName: this.firstName,
        lastName: this.lastName
      }
    })
    this.loading = false
    this.$router.back()
  }

  back (): void {
    this.$router.back()
  }
}
</script>
