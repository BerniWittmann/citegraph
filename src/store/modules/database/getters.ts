import { GetterTree } from 'vuex'
import { RootState } from '@/store/types'
import { DatabaseState } from '@/store/modules/database/types'

export const getters: GetterTree<DatabaseState, RootState> = {}
