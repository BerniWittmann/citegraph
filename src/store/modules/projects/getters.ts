import { GetterTree } from 'vuex'
import { ProjectsState } from '@/store/modules/projects/types'
import { RootState } from '@/store/types'
import Project from '@/models/project'

export const getters: GetterTree<ProjectsState, RootState> = {
  activeProject (state: ProjectsState): Project | undefined {
    return state.activeProject
  },
  hasActiveProject (state: ProjectsState): boolean {
    return !!state.activeProject
  },
  projects (state: ProjectsState): Array<Project> {
    return state.projects
  },
  openProjects (state: ProjectsState): Array<Project> {
    return state.openProjects
  }
}
