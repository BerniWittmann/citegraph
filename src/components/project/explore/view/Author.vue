<template>
  <div class="project-explore-single-author">
    <layout>
      <template v-slot:title>
        <v-avatar color="primary" class="mr-5" size="62">
          <b class="white--text">{{ initials }}</b>
        </v-avatar>
        <h2 class="display-5">{{ author.firstName }} {{ author.lastName }}</h2>
      </template>
      <v-container>
        <v-row>
          <v-col :cols="3">
            <v-list>
              <v-subheader class="title">{{ $t('project.explore.view.author.metrics') }}</v-subheader>
              <v-list-item>
                <v-list-item-content>{{ $t('project.explore.view.author.count_records') }}:</v-list-item-content>
                <v-list-item-content class="align-end">{{ author.countRecords }}</v-list-item-content>
              </v-list-item>
            </v-list>
          </v-col>
        </v-row>
      </v-container>
    </layout>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import PaperEntityLayout from '@/layouts/PaperEntity.vue'
import Author from '@/models/paperEntities/author'

@Component({
  components: {
    Layout: PaperEntityLayout
  }
})
export default class AuthorViewPage extends Vue {
  get author (): Author {
    return this.$store.getters['paperEntities/activeEntity'] as Author
  }

  get initials (): string {
    return (this.author.firstName.substring(0, 1) + this.author.lastName.substring(0, 1)).toUpperCase()
  }
}
</script>
