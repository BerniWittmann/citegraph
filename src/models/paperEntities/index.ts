import Author from './author'
import Conference from './conference'
import Country from './country'
import Institution from './institution'
import Publication from './publication'
import Publisher from './publisher'
import Record from './record'
import PaperEntity, { PaperEntityTableColumn, PaperEntityFields } from './base'

export const entities = [
  Record,
  Author,
  Institution,
  Country,
  Publication,
  Publisher,
  Conference
]

export const entityKeysMap = {
  [Record.key]: Record,
  [Author.key]: Author,
  [Institution.key]: Institution,
  [Country.key]: Country,
  [Publication.key]: Publication,
  [Publisher.key]: Publisher,
  [Conference.key]: Conference
}

export const entityKeys: Array<string> = entities.map((entity) => entity.key)

export type PaperEntityTableColumn = PaperEntityTableColumn

export type PaperEntity = PaperEntity

export type PaperEntityFields = PaperEntityFields

export default {
  Author,
  Conference,
  Country,
  Institution,
  Publication,
  Publisher,
  Record
}
