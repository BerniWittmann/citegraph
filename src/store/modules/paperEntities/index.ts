import { Module } from 'vuex'
import { PaperEntitiesState } from './types'
import { RootState } from '@/store/types'
import { state } from './state'
import { actions } from './actions'
import { getters } from './getters'
import { mutations } from './mutations'

const namespaced: boolean = true

export const paperEntities: Module<PaperEntitiesState, RootState> = {
  namespaced,
  state,
  getters,
  actions,
  mutations
}
