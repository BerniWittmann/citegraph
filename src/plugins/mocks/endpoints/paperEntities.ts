import db from '../db'
import MockAdapter from 'axios-mock-adapter'
import { RecordFields } from '@/models/paperEntities/record'
import { setupMocks as setupGraphQLMocks } from './graphql-helpers/endpoint'
import { QueryConfiguration } from './graphql-helpers/queries'
import { AuthorFields } from '@/models/paperEntities/author'
import { CountryFields } from '@/models/paperEntities/country'

const records = db.getCollection('records')
const authors = db.getCollection('authors')
const countries = db.getCollection('countries')

const queryResponseConfiguration: Record<string, QueryConfiguration> = {
  'Records': {
    schemaName: 'record',
    queryName: 'Records',
    collection: records,
    filterFunction: recordsFilter
  },
  'Authors': {
    schemaName: 'author',
    queryName: 'Authors',
    collection: authors,
    filterFunction: authorsFilter
  },
  'Countries': {
    schemaName: 'country',
    queryName: 'Countries',
    collection: countries,
    filterFunction: countriesFilter
  }
}

function recordsFilter (obj: RecordFields, filter: string) {
  return obj.title.toLowerCase().includes(filter) ||
    obj.year.toString().includes(filter) ||
    obj.keywords.map(keyword => keyword.toLowerCase()).join('').includes(filter) ||
    obj.authors.map(author => (author.firstName + ' ' + author.lastName).toLowerCase()).join(';').includes(filter)
}

function authorsFilter (obj: AuthorFields, filter: string) {
  return obj.firstName.toLowerCase().includes(filter) ||
    obj.lastName.toLowerCase().includes(filter)
}

function countriesFilter (obj: CountryFields, filter: string) {
  return obj.name.toLowerCase().includes(filter)
}

export function setupMocks (mock: MockAdapter) {
  setupGraphQLMocks(mock, queryResponseConfiguration)
}
