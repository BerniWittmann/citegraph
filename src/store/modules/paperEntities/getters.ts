import { GetterTree } from 'vuex'
import { RootState } from '@/store/types'
import { PaperEntitiesState } from './types'
import { PaperEntity } from '@/models/paperEntities'

export const getters: GetterTree<PaperEntitiesState, RootState> = {
  entities (state: PaperEntitiesState): Array<PaperEntity> {
    return state.entities
  },

  entityType (state: PaperEntitiesState): string | undefined {
    return state.entityType
  },

  entityCount (state: PaperEntitiesState): number {
    return state.entityCount
  },

  activeEntity (state: PaperEntitiesState): PaperEntity | undefined {
    return state.activeEntity
  },

  hasActiveEntity (state: PaperEntitiesState): boolean {
    return !!state.activeEntity
  }
}
