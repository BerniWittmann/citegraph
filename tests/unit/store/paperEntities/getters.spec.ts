import { getters } from '@/store/modules/paperEntities/getters'
import Author from '@/models/paperEntities/author'

describe('store/modules/paperEntities/getters', () => {
  it('provides the getters', () => {
    expect(getters).toMatchSnapshot()
  })

  describe('entities', () => {
    it('returns the stored entities', () => {
      const state = {
        entities: [
          new Author({ id: '1', firstName: 'Max', lastName: 'Mustermann', countRecords: 12 }),
          new Author({ id: '2', firstName: 'Maria', lastName: 'Maier', countRecords: 24 })
        ]
      }
      const getter = getters.entities as Function

      expect(getter(state)).toEqual(state.entities)
    })

    it('returns empty array if no entities are available', () => {
      const state = {
        entities: []
      }
      const getter = getters.entities as Function

      expect(getter(state)).toEqual([])
    })
  })

  describe('entityType', () => {
    it('returns the stored entityType', () => {
      const state = {
        entityType: 'record'
      }
      const getter = getters.entityType as Function

      expect(getter(state)).toEqual(state.entityType)
    })

    it('returns undefined if no entity type is available', () => {
      const state = {
        entityType: undefined
      }
      const getter = getters.entityType as Function

      expect(getter(state)).toEqual(undefined)
    })
  })

  describe('entityCount', () => {
    it('returns the stored entityCount', () => {
      const state = {
        entityCount: 12
      }
      const getter = getters.entityCount as Function

      expect(getter(state)).toEqual(state.entityCount)
    })

    it('returns -1 if no entity count is available', () => {
      const state = {
        entityCount: -1
      }
      const getter = getters.entityCount as Function

      expect(getter(state)).toEqual(-1)
    })
  })

  describe('activeEntity', () => {
    it('returns the stored entity', () => {
      const state = {
        activeEntity: new Author({ id: '1', firstName: 'Max', lastName: 'Mustermann', countRecords: 12 })
      }
      const getter = getters.activeEntity as Function

      expect(getter(state)).toEqual(state.activeEntity)
    })

    it('returns undefined if no entity is available', () => {
      const state = {
        activeEntity: undefined
      }
      const getter = getters.activeEntity as Function

      expect(getter(state)).toEqual(undefined)
    })
  })

  describe('hasActiveEntity', () => {
    it('returns true if there is a  stored entity', () => {
      const state = {
        activeEntity: new Author({ id: '1', firstName: 'Max', lastName: 'Mustermann', countRecords: 12 })
      }
      const getter = getters.hasActiveEntity as Function

      expect(getter(state)).toBeTruthy()
    })

    it('returns false if no entity is available', () => {
      const state = {
        activeEntity: undefined
      }
      const getter = getters.hasActiveEntity as Function

      expect(getter(state)).toBeFalsy()
    })
  })
})
