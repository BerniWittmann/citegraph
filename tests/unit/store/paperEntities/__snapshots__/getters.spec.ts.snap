// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store/modules/paperEntities/getters provides the getters 1`] = `
Object {
  "activeEntity": [Function],
  "entities": [Function],
  "entityCount": [Function],
  "entityType": [Function],
  "hasActiveEntity": [Function],
}
`;
